File: adventofcode/2023/day04/day04.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

type Card struct {
	id    int
	wnums []string
	mnums []string
}

func (c *Card) won() int {
	win := 0
	for _, mnum := range c.mnums {
		for _, wnum := range c.wnums {
			if wnum == mnum {
				win++
				break
			}
		}
	}
	return win
}

func (c *Card) score() int {
	win := c.won()
	return int(math.Pow(2.0, float64(win-1)))
}

func completed(unprocessed map[int]int) bool {
	t := 0
	for _, v := range unprocessed {
		t += v
	}
	return t == 0
}

func play(cards []Card) int {
	chash := make(map[int]Card)
	whash := make(map[int]int)
	for _, card := range cards {
		chash[card.id] = card
		whash[card.id] = card.won()
	}
	queue := []int{}
	for _, card := range cards {
		queue = append(queue, card.id)
	}
	r := 0
	for len(queue) != 0 {
		r += 1
		top := queue[0]
		tcard := chash[top]
		queue = queue[1:]
		for i := tcard.id + 1; i < tcard.id+1+whash[top]; i++ {
			queue = append(queue, i)
		}
	}
	return r
}

func parseCard(line string) Card {
	cnums := strings.Split(line, ":")
	id, _ := strconv.Atoi(strings.Trim(cnums[0][5:], " "))
	nums := strings.Split(cnums[1], "|")
	wnums := strings.Fields(nums[0])
	mnums := strings.Fields(nums[1])
	return Card{id, wnums, mnums}
}

func parseCards(lines []string) []Card {
	cards := []Card{}
	for _, line := range lines {
		card := parseCard(line)
		cards = append(cards, card)
	}
	return cards
}

func part1(cards []Card) int {
	r := 0
	for _, card := range cards {
		r += card.score()
	}
	return r
}

func part2(cards []Card) int {
	// fmt.Println(cards)
	// return 0
	return play(cards)
}

func main() {
	s := time.Now()
	lines := readInput()
	cards := parseCards(lines)
	s1 := time.Now()
	fmt.Println(part1(cards))
	e1 := time.Since(s1)
	s2 := time.Now()
	fmt.Println(part2(cards))
	e2 := time.Since(s2)
	e := time.Since(s)
	fmt.Printf("Time for part1: %s, part2: %s, total: %s\n", e1, e2, e)
}

File: adventofcode/2023/day04/day04.zig
const std = @import("std");
const time = std.time;
const print = std.debug.print;
const split = std.mem.splitScalar;
const ArrayList = std.ArrayList;
const HashMap = std.AutoHashMap;
const Allocator = std.mem.Allocator;

const Card = struct {
    id: u32,
    wnums: ArrayList(u32),
    mnums: ArrayList(u32),

    const Self = @This();

    fn parse(line: []const u8, allocator: Allocator) !Self {
        var cit = split(u8, line, ':');
        const ids = cit.next().?;
        const id = std.fmt.parseInt(u32, std.mem.trim(u8, ids[5..], " "), 10) catch 1;
        const nums = std.mem.trim(u8, cit.next().?, " ");

        var nit = split(u8, std.mem.trim(u8, nums, " "), '|');
        const wnumstr = std.mem.trim(u8, nit.next().?, " ");
        const mnumstr = std.mem.trim(u8, nit.next().?, " ");

        var wnums = ArrayList(u32).init(allocator);
        var wit = split(u8, wnumstr, ' ');
        while (wit.next()) |wnum| {
            if (std.mem.eql(u8, wnum, "")) continue;
            const num = std.fmt.parseInt(u32, std.mem.trim(u8, wnum, " "), 10) catch 0;
            try wnums.append(num);
        }

        var mnums = ArrayList(u32).init(allocator);
        var mit = split(u8, mnumstr, ' ');
        while (mit.next()) |mnum| {
            if (std.mem.eql(u8, mnum, "")) continue;
            const num = std.fmt.parseInt(u32, std.mem.trim(u8, mnum, " "), 10) catch 0;
            try mnums.append(num);
        }

        return Self{ .id = id, .wnums = wnums, .mnums = mnums };
    }

    fn won(self: Self) u32 {
        var r: u32 = 0;
        for (self.mnums.items) |mnum| {
            for (self.wnums.items) |wnum| {
                if (wnum == mnum) {
                    r += 1;
                    break;
                }
            }
        }
        return r;
    }

    fn score(self: Self) u32 {
        const win = self.won();
        if (win == 0) return 0;

        return std.math.pow(u32, 2, win - 1);
    }
};

fn play(cards: ArrayList(Card), allocator: Allocator) !u32 {
    var queue = ArrayList(u32).init(allocator);
    defer queue.deinit();
    var chash = HashMap(u32, Card).init(allocator);
    defer chash.deinit();
    var whash = HashMap(u32, u32).init(allocator);
    defer whash.deinit();
    for (cards.items) |card| {
        // print("card: {any}\n", .{card});
        try queue.append(card.id);
        try chash.put(card.id, card);
        try whash.put(card.id, card.won());
    }
    var r: u32 = 0;
    while (queue.items.len != 0) {
        r += 1;
        const top = queue.pop();
        const win = whash.get(top).?;
        const tcard = chash.get(top).?;
        for ((tcard.id + 1)..(tcard.id + 1 + win)) |i| {
            try queue.append(@intCast(i));
        }
    }
    return r;
}

pub fn main() !void {
    const s = time.microTimestamp();
    const stdin = std.io.getStdIn().reader();
    const allocator = std.heap.page_allocator;

    var cards = ArrayList(Card).init(allocator);
    defer cards.deinit();

    while (true) {
        var buf: [200]u8 = undefined;
        const line = try stdin.readUntilDelimiterOrEof(&buf, '\n');
        if (line == null) break;

        const card = try Card.parse(line.?, std.heap.page_allocator);
        try cards.append(card);
    }

    var p1: u32 = 0;
    for (cards.items) |card| {
        p1 += card.score();
    }
    print("{d}\n", .{p1});
    const p2 = try play(cards, std.heap.page_allocator);
    print("{d}\n", .{p2});

    const e = time.microTimestamp();
    print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}
